### ISSUE TITLE
generalized JSON support in stan repo

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#3142]
#### Summary:

I propose that we

1. move the JSON implementation from `stan-dev/cmdstan` to `stan-dev/stan/io` and
2. generalize to handle standard R, Python and Julia output formats for +inf, -inf, and NaN.

#### Description:

As unit tests, we should test the output for +inf, -inf, and NaN as output in popular libraries by R, Python, and Julia.

The motivation is to make it easier for interfaces like PyStan and CmdStanPy to access the code without dependencies on CmdStan or cutting and pasting the parser.

Nothing other than the parser needs to be written for this---the usage is all from the interfaces because our service layer takes in `var_context` arguments directly.    Step (1) will require an additional PR for cmdstan that can be staged after this one is done.

#### Current Version:
v2.30.0

**Comment by WardBrian** [comment]
Here's a summary of the current landscape, I believe: 

Python's stdlib `json` module outputs `NaN` and `Infinity`/`-Infinity`. Note that these are _not_ quoted strings, but Javascript atoms.

CmdStanPy currently manually changes these to `"NaN"` and `"inf"`/`"-inf"` (quotes stringd, different spelling for infinite values).

R seems to lack json support in the standard library, but cmdstanr uses `jsonlite` which outputs `"NaN"` and `"Inf"`/`"-Inf"`. It seems that cmdstanr does not manually change these. 

----
Finally, it seems like `rapidjson` (which is the C++11 library CmdStan currently uses) supports:

NaN:  `NaN`, `"NaN"`

Infinity: `Infinity`, `"Infinity"`, `"Inf"`. (**Note**: contradicting the [docs](https://mc-stan.org/docs/cmdstan-guide/json.html), it appears that `"inf"` does _not_ parse)


So, in summary, Rapidjson seems more than equipped to parse the default options in R and Python already, and cmdstanpy/the doc is wrong in how it specifies this.

**Comment by WardBrian** [comment]
Forgot to mention: Julia also does not have a stdlib json library, but the common `JSON.jl` does not support nan/infinity by default and requires the user specify how they want it serialized, so it can support anything. It seems that the [StanJulia](https://github.com/StanJulia) bindings currently don't handle this special case

**Comment by rok-cesnovar** [comment]
> Here's a summary of the current landscape, I believe:

This matches with what I wrote in https://github.com/rok-cesnovar/design-docs/blob/7fe82177e6a82d55f76ac5950d64064e5ceac283/designs/0007-json_writer.md#guide-level-explanation some time ago.

![image](https://user-images.githubusercontent.com/28476796/191975782-41fe4d48-2440-43e5-939d-4fd5cfd96d87.png)

I asked the authors of jsonlite if they would be willing to support the Javascript notation (no quotations), which would mean we can have the same notation for R & Python, but was denied.

I fully support this move to the Stan lib. JSON I/O was moved to CmdStan because the assumption at that time was that the parser is only needed for CmdStan.

**Comment by WardBrian** [comment]
@rok-cesnovar do you have a sense of how tricky this would be? I assume C++ will object to duplicate code, but if we keep the original `#ifndef`s around until after the move I suspect that it will be okay?

**Comment by rok-cesnovar** [comment]
It should be doable in 2 PRs: 
- add the json code in stan but not use it anywhere in makefiles
- Use the json lib in Stan for cmdstan and remove the json lib in cmdstan
